bundles/org.openhab.binding.teslapowerwall/src/main/java/org/openhab/binding/teslapowerwall/internal/TeslaPowerwallWebTargets.java

```diff
@@ -0,0 +1,133 @@
+    private String getTokenUri;
+    private String getOperationUri;
+    private final Logger logger = LoggerFactory.getLogger(TeslaPowerwallWebTargets.class);
+
+    public TeslaPowerwallWebTargets(String ipAddress) {
+        String baseUri = "https://" + ipAddress + "/";
+        getBatterySOEUri = baseUri + "api/system_status/soe";
+        getGridStatusUri = baseUri + "api/system_status/grid_status";
+        getSystemStatusUri = baseUri + "api/system_status";
+        getMeterAggregatesUri = baseUri + "api/meters/aggregates";
+        getTokenUri = baseUri + "api/login/Basic";
+        getOperationUri = baseUri + "api/operation";
+    }
+
+    public BatterySOE getBatterySOE(String token) throws TeslaPowerwallCommunicationException {
+        String response = invoke(getBatterySOEUri, token);
+        logger.trace("getBatterySOE response = {}", response);
+        return BatterySOE.parse(response);
+    }
+
+    public GridStatus getGridStatus(String token) throws TeslaPowerwallCommunicationException {
+        String response = invoke(getGridStatusUri, token);
+        logger.trace("getGridStatus response = {}", response);
+        return GridStatus.parse(response);
+    }
+
+    public SystemStatus getSystemStatus(String token) throws TeslaPowerwallCommunicationException {
+        String response = invoke(getSystemStatusUri, token);
+        logger.trace("getSystemStatus response = {}", response);
+        return SystemStatus.parse(response);
+    }
+
+    public MeterAggregates getMeterAggregates(String token) throws TeslaPowerwallCommunicationException {
+        String response = invoke(getMeterAggregatesUri, token);
+        logger.trace("getMeterAggregates response = {}", response);
+        return MeterAggregates.parse(response);
+    }
+
+    public Operations getOperations(String token) throws TeslaPowerwallCommunicationException {
+        String response = invoke(getOperationUri, token);
+        logger.trace("getOperations response = {}", response);
+        return Operations.parse(response);
+    }
+
+    public String getToken(String email, String password) throws TeslaPowerwallCommunicationException {
+        JsonObject jsonObject = new JsonObject();
+        jsonObject.addProperty("username", "customer");
+        jsonObject.addProperty("password", password);
+        jsonObject.addProperty("email", email);
+        jsonObject.addProperty("force_sm_off", false);
+        logger.debug("logonjson = {}", jsonObject.toString());
+        String response = invoke(getTokenUri, "POST", "application/json", jsonObject.toString());
+        JsonObject jsonResponse = JsonParser.parseString(response).getAsJsonObject();
+        String token = jsonResponse.get("token").getAsString();
+        logger.debug("Token: {}", token);
+        return token;
+    }
+
+    private String invoke(String uri, String token) throws TeslaPowerwallCommunicationException {
+        return invoke(uri, "GET", token, "");
```

Comment:
The 3rd parameter `token` is passed to the method and that method expects a string `contentType` looking at the implementation, it is used as header "Content-Type" this is usually something like 'application/text' and not a token.

Reply:
Thanks,  I've  gone back and re-read  the docs  and my implementation (whilst working) was wrong :)